hw/top_earlgrey/dv/env/seq_lib/chip_sw_base_vseq.sv

```
@@ -643,6 +653,51 @@ class chip_sw_base_vseq extends chip_base_vseq;
+  virtual task jtag_otp_program32(int addr,
+                                  bit [31:0] data,
+                                  output bit err);
+
+    bit [TL_DW-1:0] status;
+    bit [TL_DW-1:0] err_mask = 0;
+    bit idle = 0;
+    int base_addr = top_earlgrey_pkg::TOP_EARLGREY_OTP_CTRL_CORE_BASE_ADDR;
+    jtag_riscv_agent_pkg::jtag_write_csr(base_addr +
+                                         ral.otp_ctrl_core.direct_access_address.get_offset(),
+                                         p_sequencer.jtag_sequencer_h,
+                                         addr);
+
+    jtag_riscv_agent_pkg::jtag_write_csr(base_addr +
+                                         ral.otp_ctrl_core.direct_access_wdata[0].get_offset(),
+                                         p_sequencer.jtag_sequencer_h,
+                                         data[31:0]);
+
+    jtag_riscv_agent_pkg::jtag_write_csr(base_addr +
+                                         ral.otp_ctrl_core.direct_access_cmd.get_offset(),
+                                         p_sequencer.jtag_sequencer_h,
+                                         1 << ral.otp_ctrl_core.direct_access_cmd.wr.get_lsb_pos());
+
+
+    while (!idle) begin
+      jtag_riscv_agent_pkg::jtag_read_csr(base_addr +
+                                          ral.otp_ctrl_core.status.get_offset(),
+                                          p_sequencer.jtag_sequencer_h,
+                                          status);
+
+      idle = status >> (ral.otp_ctrl_core.status.dai_idle.get_lsb_pos()) & TL_DW'(1'b1);
+
+      err_mask = ~((1 << ral.otp_ctrl_core.status.dai_idle.get_lsb_pos()) |
+                   (1 << ral.otp_ctrl_core.status.check_pending.get_lsb_pos()));
+
+      `uvm_info(`gfn, $sformatf("Waiting for DAI to become idle = 1, actual: %d!", idle),
+         UVM_MEDIUM)
+
+      // If any bits other than dai_idle and check pending are set, error back.
+      err = |(status & err_mask);
```

Do we ever expect the operations above to fail (I can't think of a reason given that this is a very custom purpose-driven function)? If not, then we can just throw an error here.

---

yep, in fact it failed multiple times when i was trying to write the test.
Mostly due to user error, but it's not an unreasonable expectation.  Especially if the default otp values / partitions are ever shifted around.

---

no i think it still makes sense for the caller to handle. Since there are quite a lot of error conditions associated with any otp operation.

---

on second thought I think you were right, fixed in #15755